pkg/controller/container-runtime-config/helpers.go

```diff
@@ -753,3 +781,159 @@ func convertICSPToIDMS(icsp *apioperatorsv1alpha1.ImageContentSourcePolicy) *api
+			}
+			signedOptions = append(signedOptions, signature.PRSigstoreSignedWithRekorPublicKeyData(rekorKeyDataDec))
+		}
+	case apicfgv1alpha1.FulcioCAWithRekorRootOfTrust:
+		fulcioOptions := []signature.PRSigstoreSignedFulcioOption{}
+		fulcioDataDec, err := b64.StdEncoding.DecodeString(policy.RootOfTrust.FulcioCAWithRekor.FulcioCAData)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode fulcioCAData: %v", err)
+		}
+		fulcioOptions = append(fulcioOptions, signature.PRSigstoreSignedFulcioWithCAData(fulcioDataDec),
+			signature.PRSigstoreSignedFulcioWithOIDCIssuer(policy.RootOfTrust.FulcioCAWithRekor.FulcioSubject.OIDCIssuer),
+			signature.PRSigstoreSignedFulcioWithSubjectEmail(policy.RootOfTrust.FulcioCAWithRekor.FulcioSubject.SignedEmail))
+
+		prSigstoreSignedFulcio, err := signature.NewPRSigstoreSignedFulcio(fulcioOptions...)
+		if err != nil {
+			return nil, err
+		}
+		signedOptions = append(signedOptions, signature.PRSigstoreSignedWithFulcio(prSigstoreSignedFulcio))
+
+		rekorKeyDataDec, err := b64.StdEncoding.DecodeString(policy.RootOfTrust.FulcioCAWithRekor.RekorKeyData)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode rekorKeyData: %v", err)
+		}
+		signedOptions = append(signedOptions, signature.PRSigstoreSignedWithRekorPublicKeyData(rekorKeyDataDec))
+	}
+
+	switch policy.SignedIdentity.MatchPolicy {
+	case apicfgv1alpha1.IdentityMatchPolicyRemapIdentity:
+		identity, err := signature.NewPRMRemapIdentity(string(policy.SignedIdentity.PolicyMatchRemapIdentity.Prefix), string(policy.SignedIdentity.PolicyMatchRemapIdentity.SignedPrefix))
+		if err != nil {
+			return nil, fmt.Errorf("error getting signedIdentity for %s: %v", apicfgv1alpha1.IdentityMatchPolicyRemapIdentity, err)
+		}
+		signedIdentity = identity
+	case apicfgv1alpha1.IdentityMatchPolicyExactRepository:
+		identity, err := signature.NewPRMExactRepository(string(policy.SignedIdentity.PolicyMatchExactRepository.Repository))
+		if err != nil {
+			return nil, fmt.Errorf("error getting signedIdentity for %s: %v", apicfgv1alpha1.IdentityMatchPolicyExactRepository, err)
+		}
+		signedIdentity = identity
+	case apicfgv1alpha1.IdentityMatchPolicyMatchRepository:
+		signedIdentity = signature.NewPRMMatchRepository()
+	case apicfgv1alpha1.IdentityMatchPolicyMatchRepoDigestOrExact, "":
+		signedIdentity = signature.NewPRMMatchRepoDigestOrExact()
+	default:
+		return nil, fmt.Errorf("unknown signedIdentity match policy: %s", policy.SignedIdentity.MatchPolicy)
+	}
+
+	signedOptions = append(signedOptions, signature.PRSigstoreSignedWithSignedIdentity(signedIdentity))
+
+	if sigstorePolicyRequirement, err = signature.NewPRSigstoreSigned(signedOptions...); err != nil {
+		return nil, err
+	}
+
+	return sigstorePolicyRequirement, nil
+}
+
+func validateClusterImagePolicyWithAllowedBlockedRegistries(clusterScopePolicies map[string]signature.PolicyRequirements, allowedRegs, policyBlocked []string) error {
```

Comment:
An idle thought… is this sufficient?

- If `example.com` is blocked, but a policy exists for `example.com/repo`, that effectively unblocks that namespace
- If only `example.com` is allowed, but a policy exists for `example.invalid`, that effectively unblocks that namespace

It’s quite possible this was already discussed and designed in the enhancement.

Reply:
The enhancement only mentioned rejection if the scopes and allowed/blocked list are the same, it hasn't been discussed the case the scopes are nested inside.

Reply:
@mtrmac  What are your thoughts on this?
- if blockedlist exists, the clusterimagepolicy scopes must not equal to or nested under blockedRegistries
- if allowedlist exists, the clusterimagepolicy scopes nested under the allowedRegistries

Reply:
I don’t really have a strong opinion here — on the one hand the behavior in the current version of the code is _pretty surprising_, OTOH handling conflicts by ~ignoring configuration and MCO just reporting something in a status is probably also not very intuitive, and minimizing the cases where configuration is ignored that way seems valuable.

But it also feels like an instance of a general design concern where the k8s universe should have a general best practice. I wouldn’t know what that is.

Reply:
(I didn’t see your last comment before)

> * if blockedlist exists, the clusterimagepolicy scopes must not equal to or nested under blockedRegistries
> * if allowedlist exists, the clusterimagepolicy scopes nested under the allowedRegistries

I agree this makes sense as a consistent configuration.

What happens when we go from a consistent configuration to an inconsistent one? E.g. supporting (allowed=A; CIP scopes = [A, A/repo]), and then an admin adds a new CIP scope B.

Is the effect that we retain the current constant `MachineConfig`s, or would the code fall open and ignore everything?

Reply:
> Is the effect that we retain the current constant MachineConfigs, or would the code fall open and ignore everything?

retain the current constant MachineConfigs, the controller would fail to sync the new inconsistent CIP.

Reply:
Great :) In that case the extra checks proposed in https://github.com/openshift/machine-config-operator/pull/4160#discussion_r1483590046 look good to me.